File: src/engagement/seasons.js
/**
 * Seasons (gamification spec §12). The season follows the device calendar by
 * quarter: Mar–May spring, Jun–Aug summer, Sep–Nov autumn, Dec–Feb winter.
 *
 * This module starts minimal for the store's availability rules (§08 —
 * seasonal birds can be given a home only while they are here, and the store
 * says "back next summer", never a countdown). The ceremonies step extends it
 * with tints, particles, night and the leaving/return lifecycle.
 */

export const SEASONS = ["spring", "summer", "autumn", "winter"];

export const SEASON_META = {
  spring: { id: "spring", name: "Spring", title: "First Light" },
  summer: { id: "summer", name: "Summer", title: "Long Days" },
  autumn: { id: "autumn", name: "Autumn", title: "The Leaving" },
  winter: { id: "winter", name: "Winter", title: "Quiet Meadow" },
};

export function seasonForDate(date = new Date()) {
  const m = date.getMonth();
  if (m >= 2 && m <= 4) return "spring";
  if (m >= 5 && m <= 7) return "summer";
  if (m >= 8 && m <= 10) return "autumn";
  return "winter";
}

/** "back next summer" — the season a visitor returns in, kindly worded. */
export function returnLine(species) {
  if (!species.seasons?.length) return null;
  return `back next ${species.seasons[0]}`;
}

/** "Summer only" / "Winter only" chips for the store. */
export function seasonChip(species) {
  if (!species.seasons?.length) return null;
  if (species.seasons.length === 1) return `${SEASON_META[species.seasons[0]].name} only`;
  return `${SEASON_META[species.seasons[0]].name}–${SEASON_META[species.seasons[species.seasons.length - 1]].name}`;
}

// --- §12: a season changes exactly three things — the canopy and ground
// tint, one drifting particle, and which visitor is in the store. Birds and
// interface never retint. ---

export const SEASON_TINTS = {
  spring: { canopy: "#8FD9C8", ground: "#7FCFBE", sky: "#C9E8DF", particle: "petals" },
  summer: { canopy: "#3E9E8E", ground: "#6FC3B2", sky: "#C9E8DF", particle: "seeds" },
  autumn: { canopy: "#F26B3A", ground: "#F0A47A", sky: "#C9E8DF", particle: "leaves" },
  winter: { canopy: "#B9CFC9", ground: "#FFFBEB", sky: "#DCEDF2", particle: "snow" },
};

/** Night is a palette swap after 7pm — 9pm in summer (§12). */
export function isNight(date = new Date(), season = seasonForDate(date)) {
  const h = date.getHours();
  const nightAt = season === "summer" ? 21 : 19;
  return h >= nightAt || h < 6;
}

/** "2026-autumn" — the key departure/return events are remembered under. */
export function seasonKeyForDate(date = new Date()) {
  return `${date.getFullYear()}-${seasonForDate(date)}`;
}

function previousSeason(season) {
  return SEASONS[(SEASONS.indexOf(season) + 3) % 4];
}

function nextSeason(season) {
  return SEASONS[(SEASONS.indexOf(season) + 1) % 4];
}

/** §11: a week of restlessness — present now, gone next season. */
export function leavingSoon(species, date = new Date()) {
  if (!species.seasons) return false;
  const season = seasonForDate(date);
  const monthEnd = [4, 7, 10, 1][SEASONS.indexOf(season)];
  const boundary = new Date(
    date.getFullYear() + (season === "winter" && date.getMonth() === 11 ? 1 : 0),
    monthEnd + 1,
    1
  );
  const daysLeft = Math.ceil((boundary - date) / 86400000);
  return (
    daysLeft <= 7 &&
    species.seasons.includes(season) &&
    !species.seasons.includes(nextSeason(season))
  );
}

/**
 * §11 events on Meadow open, derived from the calendar + seen-marks.
 * A departure is due when an OWNED migrant was here last season and is gone
 * now, and this season's send-off hasn't played (or been missed). A return is
 * due when an owned migrant is back and this season's arrival hasn't played.
 * Away birds keep their perch and count toward everything.
 */
export function migrationEvents(state, ownedSpecies, date = new Date()) {
  const season = seasonForDate(date);
  const key = seasonKeyForDate(date);
  const departures = [];
  const returns = [];
  for (const species of ownedSpecies) {
    if (!species.seasons) continue;
    const here = species.seasons.includes(season);
    const wasHere = species.seasons.includes(previousSeason(season));
    if (!here && wasHere && state.departuresSeen?.[species.id] !== key) departures.push(species);
    if (here && !wasHere && state.returnsSeen?.[species.id] !== key) returns.push(species);
  }
  return { departures, returns, season, key };
}

export function isAway(species, date = new Date()) {
  return Boolean(species.seasons) && !species.seasons.includes(seasonForDate(date));
}

/** "Away · back in spring" — the chip on an away bird's entry (§11). */
export function awayChip(species) {
  if (!species.seasons?.length) return null;
  return `Away · back in ${species.seasons[0]}`;
}

File: ios/KidMath/Services/Seasons.swift
import SwiftUI

/// Seasons (§12) — Swift mirror of src/engagement/seasons.js. The season
/// follows the device calendar by quarter and changes exactly three things:
/// the canopy/ground tint, one particle, and which visitor is in the store.
/// Night is a palette swap after 7pm (9pm in summer); birds and interface
/// never retint.
enum Seasons {

    nonisolated static func current(_ date: Date = Date()) -> String {
        switch Calendar.current.component(.month, from: date) {
        case 3...5: return "spring"
        case 6...8: return "summer"
        case 9...11: return "autumn"
        default: return "winter"
        }
    }

    nonisolated static func isNight(_ date: Date = Date()) -> Bool {
        let hour = Calendar.current.component(.hour, from: date)
        let nightAt = current(date) == "summer" ? 21 : 19
        return hour >= nightAt || hour < 6
    }

    /// Species presence from the roster's `seasons` list (nil = resident).
    nonisolated static func presentNow(_ speciesRaw: [String: Any], date: Date = Date()) -> Bool {
        guard let seasons = speciesRaw["seasons"] as? [String] else { return true }
        return seasons.contains(current(date))
    }

    /// "Away · back in winter" — the kind chip on an away bird (§11).
    nonisolated static func awayLine(_ speciesRaw: [String: Any]) -> String? {
        guard let seasons = speciesRaw["seasons"] as? [String], let first = seasons.first else { return nil }
        return "Away · back in \(first)"
    }

    nonisolated static let order = ["spring", "summer", "autumn", "winter"]

    nonisolated static func previous(_ season: String) -> String {
        let index = order.firstIndex(of: season) ?? 0
        return order[(index + 3) % 4]
    }

    /// "2026-autumn" — the key departure/return events are remembered under
    /// (mirror of seasonKeyForDate in src/engagement/seasons.js).
    nonisolated static func seasonKey(_ date: Date = Date()) -> String {
        "\(Calendar.current.component(.year, from: date))-\(current(date))"
    }

    struct MigrationEvents {
        let departures: [String]
        let returns: [String]
        let key: String
    }

    /// §11 events on Meadow open, derived from the calendar + seen-marks: a
    /// departure is due when an OWNED migrant was here last season and is
    /// gone now (send-off unseen); a return when she is back and the arrival
    /// hasn't played. Away birds keep their perch and count toward everything.
    nonisolated static func migrationEvents(
        state: [String: Any],
        ownedSpecies: [(id: String, raw: [String: Any])],
        date: Date = Date()
    ) -> MigrationEvents {
        let season = current(date)
        let key = seasonKey(date)
        let departuresSeen = state["departuresSeen"] as? [String: Any] ?? [:]
        let returnsSeen = state["returnsSeen"] as? [String: Any] ?? [:]
        var departures: [String] = []
        var returns: [String] = []
        for species in ownedSpecies {
            guard let seasons = species.raw["seasons"] as? [String] else { continue }
            let here = seasons.contains(season)
            let wasHere = seasons.contains(previous(season))
            if !here, wasHere, departuresSeen[species.id] as? String != key { departures.append(species.id) }
            if here, !wasHere, returnsSeen[species.id] as? String != key { returns.append(species.id) }
        }
        return MigrationEvents(departures: departures, returns: returns, key: key)
    }

    /// Canopy/ground tints per season (the §12 hexes). `nil` season (flag
    /// off) keeps the base palette.
    struct Tint {
        let sky: Color
        let farGrass: Color
        let ground: Color
        let groundDeep: Color
        let canopy: Color
        let frozen: Bool
    }

    nonisolated static func tint(for season: String?) -> Tint {
        switch season {
        case "spring":
            return Tint(
                sky: Color(red: 0.79, green: 0.91, blue: 0.87),
                farGrass: Color(hexValue: 0x7FCFBE),
                ground: Color(hexValue: 0x7FCFBE).opacity(0.85),
                groundDeep: Color(hexValue: 0x8FD9C8),
                canopy: Color(hexValue: 0x8FD9C8),
                frozen: false
            )
        case "summer":
            return Tint(
                sky: Color(red: 0.79, green: 0.91, blue: 0.87),
                farGrass: Color(hexValue: 0x6FC3B2),
                ground: Color(hexValue: 0x6FC3B2).opacity(0.85),
                groundDeep: Color(hexValue: 0x3E9E8E),
                canopy: Color(hexValue: 0x3E9E8E),
                frozen: false
            )
        case "autumn":
            return Tint(
                sky: Color(red: 0.79, green: 0.91, blue: 0.87),
                farGrass: Color(hexValue: 0xF0A47A),
                ground: Color(hexValue: 0xF0A47A).opacity(0.85),
                groundDeep: Color(hexValue: 0xF26B3A),
                canopy: Color(hexValue: 0xF26B3A),
                frozen: false
            )
        case "winter":
            return Tint(
                sky: Color(hexValue: 0xDCEDF2),
                farGrass: Color(hexValue: 0xD8E5DF),
                ground: Color(hexValue: 0xFFFBEB),
                groundDeep: Color(hexValue: 0xCBDDD6),
                canopy: Color(hexValue: 0xB9CFC9),
                frozen: true
            )
        default:
            return Tint(
                sky: Color(red: 0.79, green: 0.91, blue: 0.87),
                farGrass: Color(red: 0.56, green: 0.82, blue: 0.75),
                ground: Color(red: 0.65, green: 0.87, blue: 0.83),
                groundDeep: Color(red: 0.5, green: 0.81, blue: 0.75),
                canopy: Color(red: 0.44, green: 0.76, blue: 0.7),
                frozen: false
            )
        }
    }
}

private extension Color {
    init(hexValue: UInt32) {
        self.init(
            red: Double((hexValue >> 16) & 0xFF) / 255,
            green: Double((hexValue >> 8) & 0xFF) / 255,
            blue: Double(hexValue & 0xFF) / 255
        )
    }
}
